File: grind75-01-two-sum.go
// Package twosum https://leetcode.com/problems/two-sum
package twosum

// O(n) solution using maps - 70th percentile
func twoSum(nums []int, target int) []int {
	oldVals := map[int]int{} // { oldVal: oldIndex }
	for i, num := range nums {
		complement := target - num
		oldIndex, ok := oldVals[complement]
		if ok {
			return []int{i, oldIndex}
		}
		oldVals[num] = i
	}
	return []int{}
}

File: grind75-01-two-sum.ts
// https://leetcode.com/problems/two-sum
// O(n) solution using obj's - 80th percentile
function twoSum(nums: number[], target: number): number[] {
    let oldVals = {}  // { oldVal: oldIndex }
    for (let [i, num] of nums.entries()) {
        let complement = target - num
        let oldIndex = oldVals[complement]
        if (oldIndex !== undefined) {
            return [oldIndex, i]
        }
        oldVals[num] = i
    }
    return []
}
